src/components/tables/Table.tsx

```diff
@@ -0,0 +1,173 @@
+  actionBar?: true | IActionBarProps;
+  getActions?: (rowData?: ITableRowData, props?: ITableProps) => IActionOptions[];
+  blankSlateNoResultsOnAction?: IBlankSlateProps;
+  blankSlateOnError?: IBlankSlateProps;
+  filter?: true | IFilterBoxProps;
+  filterMethod?: (attributeValue: any, props: ITableOwnProps) => boolean;
+  predicates?: ITablePredicate[];
+  navigation?: true | INavigationChildrenProps;
+  lastUpdatedLabel?: string;
+  manual?: (
+    tableOwnProps: ITableOwnProps,
+    shouldResetPage: boolean,
+    tableCompositeState: ITableCompositeState,
+    previousTableCompositeState: ITableCompositeState,
+  ) => IThunkAction;
+};
+
+export interface ItableCompositeStateProps {
+  readonly tableCompositeState?: ITableCompositeState;
+};
+
+export interface ITableProps extends ITableOwnProps, ItableCompositeStateProps, Partial<ITableDispatchProps> { }
+
+export class Table extends React.Component<ITableProps, any> {
+  private updateCountForLoadingBehavior: number = 0;
+  private isInitialLoad: boolean;
+
+  static defaultProps = {
+    tableCompositeState: { sortState: {} } as any,
+    initialTableData: DEFAULT_TABLE_DATA,
+  } as Partial<ITableOwnProps>;
+
+  constructor(props: ITableProps) {
+    super(props);
+
+    // Only use the initial load strategy for tables that do not provide initialTableData in their own props
+    this.isInitialLoad = props.initialTableData == DEFAULT_TABLE_DATA;
+  }
+
+  componentDidMount() {
+    if (this.props.onDidMount) {
+      this.props.onDidMount();
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.updateCountForLoadingBehavior < 2) {
+      this.updateCountForLoadingBehavior += 1;
+    } else {
+      /**
+       *  A count is utilized for a clean loading behavior during the initial load of the table
+       *  The first count occurs after mount
+       *  The second count occurs before the real data has loaded in the table
+       *  The initial load is than completed on the third update after the real data has loaded
+       */
+      this.isInitialLoad = false;
+    }
+  }
+
+  componentWillReceiveProps(nextProps: ITableProps) {
+    const { tableCompositeState } = this.props;
+
+    if (this.hasTableCompositeStateChanged(tableCompositeState, nextProps.tableCompositeState)) {
+      // if the change occurs outside the navigation (per page, pagination), reset the pagination to 0
+      const shouldResetPage = tableCompositeState.page === nextProps.tableCompositeState.page
+        && tableCompositeState.perPage === nextProps.tableCompositeState.perPage;
+
+      this.props.onModifyData(shouldResetPage, nextProps.tableCompositeState, tableCompositeState);
+    }
+  }
+
+  componentWillUnmount() {
+    if (this.props.onUnmount) {
+      this.props.onUnmount();
+    }
+  }
+
+  private hasTableCompositeStateChanged(currentTableCompositeState: ITableCompositeState, nextTableCompositeState: ITableCompositeState): boolean {
+    return !!currentTableCompositeState && (
```

Comment:
3 points:

1. Je sortirais la petite fonction pour checker si les predicates on changé, ça aiderait juste à voir ce que ça fait en 1 coup d'oeil, ça ferait une moins grosse charge...

2. Je remplacerais les isEmpty par des comparaison de lenght. Enfait je suis même pas de comprendre pourquoi on fait un isEmpty... Mais d'après moi comparer la length devrait être suffisant.. Corrige-moi si je me trompe :)

3. Je sais pas si c'est mieux, mais tu pourrais checker voir si ça serait pas plus claire de mettre la chaine de OR dans un some. Je te fais un exemple des 3 modifications applique les si tu crois que c'est plus claire :)

```
const predicatesChanged =  _.some(currentTableCompositeState.predicates,
    (attributeValue: any, attributeName: string) => attributeValue !== nextTableCompositeState.predicates[attributeName],
);
return !!currentTableCompositeState && _.some([
    currentTableCompositeState.filter !== nextTableCompositeState.filter,
    currentTableCompositeState.perPage !== nextTableCompositeState.perPage,
    currentTableCompositeState.page !== nextTableCompositeState.page,
    currentTableCompositeState.sortState.attribute !== nextTableCompositeState.sortState.attribute,
    currentTableCompositeState.sortState.order !== nextTableCompositeState.sortState.order,
    currentTableCompositeState.predicates.length !== nextTableCompositeState.predicates.length,
    predicatesChanged,
], (changed: boolean) => changed);
```

Reply:
i could refactor it in many ways but i think the current implementation is good enough, we can talk about it live if you want